go/oasis-node/cmd/debug/txsource/txsource.go

```diff
@@ -0,0 +1,109 @@
+package txsource
+
+import (
+	"context"
+	"crypto"
+	"fmt"
+	"math/rand"
+
+	"github.com/spf13/cobra"
+	flag "github.com/spf13/pflag"
+	"github.com/spf13/viper"
+
+	"github.com/oasislabs/oasis-core/go/common/crypto/drbg"
+	"github.com/oasislabs/oasis-core/go/common/crypto/mathrand"
+	"github.com/oasislabs/oasis-core/go/common/logging"
+	consensus "github.com/oasislabs/oasis-core/go/consensus/api"
+	"github.com/oasislabs/oasis-core/go/control/api"
+	genesisFile "github.com/oasislabs/oasis-core/go/genesis/file"
+	"github.com/oasislabs/oasis-core/go/oasis-node/cmd/common"
+	cmdFlags "github.com/oasislabs/oasis-core/go/oasis-node/cmd/common/flags"
+	cmdGrpc "github.com/oasislabs/oasis-core/go/oasis-node/cmd/common/grpc"
+	runtimeClient "github.com/oasislabs/oasis-core/go/runtime/client/api"
+)
+
+const CfgSeed = "seed"
+
+var (
+	logger      = logging.GetLogger("cmd/txsource")
+	txsourceCmd = &cobra.Command{
+		Use:   "txsource",
+		Short: "send random transactions",
+		RunE:  doRun,
+	}
+)
+
+func doRun(cmd *cobra.Command, args []string) error {
+	cmd.SilenceUsage = true
+
+	if err := common.Init(); err != nil {
+		common.EarlyLogAndExit(err)
+	}
+
+	// Set up the genesis system for the signature system's chain context.
+	genesis, err := genesisFile.DefaultFileProvider()
+	if err != nil {
+		return fmt.Errorf("genesisFile.DefaultFileProvider: %w", err)
+	}
+	genesisDoc, err := genesis.GetGenesisDocument()
+	if err != nil {
+		return fmt.Errorf("genesis.GetGenesisDocument: %w", err)
+	}
+	logger.Debug("setting chain context", "chain_context", genesisDoc.ChainContext())
+	genesisDoc.SetChainContext()
+
+	// Set up the deterministic random source.
+	hash := crypto.SHA512
+	seed := []byte(viper.GetString(CfgSeed))
+	src, err := drbg.New(hash, seed, nil, []byte("txsource workload generator v1"))
+	if err != nil {
+		return fmt.Errorf("drbg.New: %w", err)
+	}
+	rng := rand.New(mathrand.New(src))
+
+	// Set up the gRPC client.
+	logger.Debug("dialing node", "addr", viper.GetString(cmdGrpc.CfgAddress))
+	conn, err := cmdGrpc.NewClient(cmd)
+	if err != nil {
+		return fmt.Errorf("cmdGrpc.NewClient: %w", err)
+	}
+	defer conn.Close()
+
+	// Set up the consensus client.
+	cnsc := consensus.NewConsensusClient(conn)
+
+	// Set up the runtime client.
+	rtc := runtimeClient.NewRuntimeClient(conn)
+
+	// Wait for sync before transferring control to the workload.
+	ncc := api.NewNodeControllerClient(conn)
+	logger.Debug("waiting for node sync")
+	if err = ncc.WaitSync(context.Background()); err != nil {
+		return fmt.Errorf("node controller client WaitSync: %w", err)
+	}
+	logger.Debug("node synced")
+
+	logger.Debug("entering workload")
```

Comment:
Even though this currently only has a single workload it would be useful if this could run multiple workloads at once -- e.g., similar as we have for E2E scenarios, have a way to register workloads and then start them in parallel.

Reply:
e2e scenarios run their own network. the txsource doesn't (although there is an e2e scenario that runs a network and a txsource). we could use a process manager to run multiple txsource programs in parallel, so let's not roll that ourselves. it would complicate the configuration of any one workload (currently only the seed).

Reply:
although I should add the workload name to the rng initialization. that way we can have multiple workloads use the same seed without generating the same accounts

Reply:
Even if we only support running one workload per process, there should still be a way to specify the name of the workload to run? And this would benefit from a generic way to register workloads and call them based on their name.

Reply:
that's added in 'split out workloads'

Reply:
Oh, I see, looks good, thanks!